### PR TITLE
Rework farmer field

### LINKED ISSUE #1093: Farmer and Field interaction could use an overhaul
Things I noticed:
- Fields use the citizens names, instead of UUID's
- Field update packets are sent way too often
- Farmer/Field serialization could be improved
- Farmer error checking is seriously lacking, which is why we need to call "sync" methods to detect missing fields.

### LINKED ISSUE #1857: Multiple Farmers -  Assignment issue
mine craft 1.12.2
mine colonies 8.5989

Once you have a mine colony with 2 farms.   I have 5 fields.  I try to set them up to "manual".  I have one farmer hut level 4.  I have a new farmer hut lvl 1. 

Lvl 4 hut is assign f farms 1 - 4 and lvl 1 hut assigned farm 5.

If you go to the fields and right click on the scarecrow.  It has farmer 1 (lvl 4) is assigned to every scarecrow.    Even though the farmer hut shows it correctly just farms 1 - 4.    

It doesn't appear that you can get it done correctly.   If you remove all assignments from each hut  and break the scarecrow and reassign the farmers back to fields.    When it finally updates,  it shows farmer (lvl4 hut) is assigned to all the fields.

### LINKED ISSUE #1868: farmer assignment - one field assigned to 2 farmers
minecraft 1.12.2
minecolonie 8.6037

2 farmers,  set to manual assignment.

I destroyed every scarecrow.  Then saved and quit.  Then reloaded.   Then went into the gui.   It showed no fields.   So I created new scarecrow and placed them.   it still showed no fields assigned when right click on scarecrow.  A OK .

Then went to farmer 2 and click on "Barley" field to manually field.   Went to field, it showed farmer 2?  A OK
Then went to farmer 1 and went to the assignment gui.   EVERY FIELD showed up and was able to assign the field to farmer 1 also.  So I can assign "Barley" field to farmer 1 even though it is already assigned to farmer 2.

The gui should reflect that fields as disabled  like if you have it in auto mode.  You should not be able to assign a field to another farmer when it is already assigned.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #1868
Closes #1857
Closes #1093

# Changes proposed in this pull request:
- Rewrites the farmer field completely in a way cleaner fashion.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/network/messages/OpenInventoryMessage.java [hunk 1]
```diff
@@ -186,7 +185,7 @@ private static void doFieldInventory(final OpenInventoryMessage message, final E
     {
         if (checkPermissions(ColonyManager.getClosestColony(player.getEntityWorld(), message.tePos), player))
         {
-            @NotNull final InventoryField inventoryField = ColonyManager.getColony(message.colonyId).getField(message.tePos).getInventoryField();
+            //@NotNull final InventoryField inventoryField = ColonyManager.getColony(message.colonyId).getField(message.tePos).getInventoryField();
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/network/messages/OpenInventoryMessage.java:188]: MINOR This block of commented-out lines of code should be removed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: src/main/java/com/minecolonies/coremod/network/messages/OpenInventoryMessage.java [hunk 2]
```diff
@@ -186,7 +185,7 @@ private static void doFieldInventory(final OpenInventoryMessage message, final E
...
             if (!StringUtils.isNullOrEmpty(message.name))
             {
                 // inventoryField.setCustomName(message.name);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/network/messages/OpenInventoryMessage.java:191]: MINOR This block of commented-out lines of code should be removed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: src/main/java/com/minecolonies/coremod/colony/Colony.java [hunk 3]
```diff
@@ -227,6 +228,7 @@
      * @param w  The world the colony exists in.
      * @param c  The center of the colony (location of Town Hall).
      */
+    @SuppressWarnings("squid:S2637")
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/Colony.java:231]: INFO Suppressing the 'squid:S2637' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 72 issues
* ![MAJOR][MAJOR] 16 major
* ![MINOR][MINOR] 49 minor
* ![INFO][INFO] 7 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [WindowHutFarmer.java#L214](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/client/gui/WindowHutFarmer.java#L214): Reduce this anonymous class number of lines from 43 to at most 40, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [Colony.java#](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#): This file has 1,376 lines, which is greater than 1,000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L59](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#L59): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 50 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L59](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#L59): class "Colony" has 101 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L353](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#L353): Remove this useless assignment to local variable "achievementKey". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L536](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#L536): Remove this useless assignment to local variable "achievementCompound". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L658](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#L658): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L1106](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1106): The Cyclomatic Complexity of this method "onWorldTick" is 20 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [Colony.java#L1622](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1622): Remove this useless assignment to local variable "size". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L315](https://github.com/Minecolonies/minecolonies/blob/48b26079ad26e29ac3f608fc0718cb44cecb9957/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L315): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — DISMISSED** [review]

File: src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java [hunk 4]
```diff
@@ -1,70 +1,129 @@
...
+import net.minecraft.init.Blocks;
+import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
 import net.minecraft.network.NetworkManager;
 import net.minecraft.network.play.server.SPacketUpdateTileEntity;
 import net.minecraft.tileentity.TileEntityChest;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.IWorldNameable;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java:20]: MINOR Remove this unused import 'net.minecraft.world.IWorldNameable'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 73 issues
* ![MAJOR][MAJOR] 16 major
* ![MINOR][MINOR] 50 minor
* ![INFO][INFO] 7 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [WindowHutFarmer.java#L214](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/client/gui/WindowHutFarmer.java#L214): Reduce this anonymous class number of lines from 43 to at most 40, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [Colony.java#](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#): This file has 1,376 lines, which is greater than 1,000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L59](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L59): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 50 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L59](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L59): class "Colony" has 101 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L353](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L353): Remove this useless assignment to local variable "achievementKey". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L536](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L536): Remove this useless assignment to local variable "achievementCompound". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L658](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L658): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L1106](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1106): The Cyclomatic Complexity of this method "onWorldTick" is 20 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [Colony.java#L1622](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1622): Remove this useless assignment to local variable "size". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L315](https://github.com/Minecolonies/minecolonies/blob/57b954df6812a7ac2d1793f39415eee58f08f0a0/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L315): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**